### DISCUSSION TITLE
Water Flow Sensor for Tasmota

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#15302] [Show and tell]
This is a pre-release copy of my water Flow Sensor for TASMOTA...
I have been using it for a few weeks now and it works well in my system feeding data to Grafana and NodeRed.

I started the development using the "xsns_01_counter.ino" as my guide.
I'm open to suggestions on what I may need to change or what I should modify in this code.

Detailed information is here:
https://github.com/trlafleur/Flow-Sensor-Tasmota

**Comment by sfromis** [comment]
Only detail available clicking on that link is `404` ;-)

**Reply by trlafleur** [reply]
Ops, it's now fixed...

**Comment by barbudor** [comment]
Hi @trlafleur 

I had a short look at your code, not from the functional point of view but globally as Tasmota Driver good rules of implementation

- I believe that it is unlikely that 40 bytes find their way to the flash Settings. I think that you will need a boot-rule to setup your configuration on each restart. autoexec.bat is a also good option when you have a file system. Alternative, for a driver that would in all case need custom compile, is overridable #defines to set the initial value. Those can then be tweaked in the console and once finalized, hard-coded back as overriders initial values. Of course the boot rule is more convenient as you can change the values without recompiling.
- All symbols (variables and functions) should have a name that is 100% unique across all Tasmota code. MySettings, MyBuf, current_pulse_counter, BoundCheck() ... should preferably be prepended by something that is unique to your driver such as `FlowCtr_...`
- Your driver is using too much static RAM. All globals should be moved into a structure (like `FlowCtr_Data`) and you should only have as global variable a pointer to an instance of that structure `FlowCtr_Data *FlowCtr_data = nullptr` that is allocated only if the driver is required during the `FlowInit()`. You can then use `if (FlowCtr_data) ...` instead of `if (Flow.counter) ...` to check thtat the driver has been initialized.
The only variables that may have to remains outside of that structure are the 3 volatiles that interface with the ISR. I'm not sure that the compiler could honor a volatile field inside a structure.
- Note that the `volatile` on the 2 variables inside the ISR is useless (and probably counter-productive) as those 2 variables are allocated on the stack only for the duration of the ISR call. It's better to let the compiler optimize them.

Thanks

**Comment by trlafleur** [comment]
Thanks for the update, I will look at making changes to implement your thoughts... as there is no document on building a driver, I was somewhat in the dark on how to develop this... but everything makes sense.

Are there any examples of using boot-rule or autoexec.bat to save and restore settings structures??

The global Settings structure, as you pointed out is almost full... what are the plans if any, to solve this issue as Tasmota keeps growing...

volatile in the ISR was an oversight when I move them from a global to the ISR

Thanks for your help...

**Comment by trlafleur** [comment]
I have implemented most of your suggestions, but when I run the code with dynamic allocation of the struct, the system crashes on startup...  can't seem to resolve this...

~~~
struct FLOWCTR
{
  uint8_t     no_pullup;                 // Counter input pullup flag (1 = No pullup)
  uint8_t     pin_state;                 // LSB0..3 Last state of counter pin; LSB7==0 IRQ is FALLING, LSB7==1 IRQ is CHANGE
  uint32_t    timer;                     // Last flow time in micro seconds
  uint32_t    timer_low_high;            // Last low/high flow counter time in micro seconds
  bool        counter;                   // 
  uint32_t    OneHour;                   // event timers....
  uint32_t    OneDay;
  uint32_t    SendingRate;               // Current sample rate count
  uint32_t    LastPulseCount;            // use to check for new flow
  uint32_t    CurrentTime;               // Current time 
  uint32_t    LastPulseTime;             // Last pulse time
  uint32_t    CurrentFlowStartTime;      // Start of current flow      
  uint32_t    ExcessiveFlowStartTime;    // Time that this flow started...
  float       Saved1hrFlowVolume;        // saved volume for the last 1hr period
  float       Saved24hrFlowVolume;       // saved volume for the last 24hr period
  float       Current1hrFlowVolume;      // current run-rate volume for this 1hr period
  float       Current24hrFlowVolume;     // current run-rate volume for this 24hr period
  float       VolumePerFlow;             // Volume for current flow
  float       CurrentFlow;
  float       OldFlowRate;
  char        Current_Units[8];           // Default units...
  char        Current_Volume_Units[8];
  bool        WeHaveFlow;                 // true if we have started a new flow
  bool        WeHaveFlowOverThreshold;    // true if current flow exceed threshold
  bool        WeHaveExcessFlow;           // true if we have exceed threshold and we have exceeded flow_threshold_time
} ;
struct FLOWCTR *FlowCtr;

FlowCtr = (struct FLOWCTR*) malloc(sizeof(struct FLOWCTR));
~~~

**Reply by barbudor** [reply]
The malloc should be done in the Init() function
I'm not at my PC right now but récent drivers are implemented this
I did that in https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_93_hm330x.ino

**Reply by barbudor** [reply]
The point is doing the malloc only if you need the driver (ie the user has defined a flow counter gpio)

**Comment by trlafleur** [comment]
~~~
Linking .pio/build/tasmota32/firmware.elf
Retrieving maximum program size .pio/build/tasmota32/firmware.elf
Checking size .pio/build/tasmota32/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  19.1% (used 62628 bytes from 327680 bytes)
Flash: [=======   ]  74.0% (used 1407129 bytes from 1900544 bytes)
Building .pio/build/tasmota32/firmware.bin
esptool.py v3.2.1
Creating esp32 image...
Merged 25 ELF sections
Successfully created esp32 image.
esp32_create_combined_bin([".pio/build/tasmota32/firmware.bin"], [".pio/build/tasmota32/firmware.elf"])
    Offset | File
 -  0x1000 | /Users/lafleur/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32/bin/bootloader_dout_40m.bin
 -  0x8000 | /Users/lafleur/Desktop/Tasmota11/.pio/build/tasmota32/partitions.bin
 -  0xe000 | /Users/lafleur/.platformio/packages/framework-arduinoespressif32/tools/partitions/boot_app0.bin
 - 0x10000 | /Users/lafleur/Desktop/Tasmota11/.pio/build/tasmota32/firmware.bin
esptool.py v3.2.1
Wrote 0x168f90 bytes to file /Users/lafleur/Desktop/Tasmota11/.pio/build/tasmota32/firmware.factory.bin, ready to flash to offset 0x0
bin_map_copy([".pio/build/tasmota32/firmware.bin"], [".pio/build/tasmota32/firmware.elf"])
map_gzip([".pio/build/tasmota32/firmware.bin"], [".pio/build/tasmota32/firmware.elf"])
======================================================================================================== [SUCCESS] Took 45.43 seconds ========================================================================================================

Environment    Status    Duration
-------------  --------  ------------
tasmota32      SUCCESS   00:00:45.426
~~~
~~~
[15:57:04]Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
[15:57:04]
[15:57:04]Core  1 register dump:
[15:57:04]PC      : 0x4011498f  PS      : 0x00060c30  A0      : 0x8010497e  A1      : 0x3ffb2700  
[15:57:04]A2      : 0x5dbac26c  A3      : 0x3ffc8784  A4      : 0x00000002  A5      : 0x00000000  
[15:57:04]A6      : 0x3ffc8200  A7      : 0x3ffb2700  A8      : 0x800e2bea  A9      : 0x3ffb2730  
[15:57:04]A10     : 0x00000045  A11     : 0x00000000  A12     : 0x00000045  A13     : 0x00003fc0  
[15:57:04]A14     : 0x00000001  A15     : 0x3ffb08c0  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c  
[15:57:04]EXCVADDR: 0x0000000c  LBEG    : 0x400870fd  LEND    : 0x40087105  LCOUNT  : 0x00000027  
[15:57:04]
[15:57:04]
[15:57:04]Backtrace:0x4011498c:0x3ffb27000x4010497b:0x3ffb2770 0x40150351:0x3ffb2790 0x400e601e:0x3ffb2820 
[15:57:04]
[15:57:04]
[15:57:04]
[15:57:04]
[15:57:04]ELF file SHA256: 0000000000000000
[15:57:04]
[15:57:04]Rebooting...
[15:57:04]ets Jun  8 2016 00:22:57
[15:57:04]
[15:57:04]rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
[15:57:04]configsip: 0, SPIWP:0xee
[15:57:04]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[15:57:04]mode:DOUT, clock div:2
[15:57:04]load:0x3fff0018,len:4
[15:57:04]load:0x3fff001c,len:1044
[15:57:04]load:0x40078000,len:10124
[15:57:04]load:0x40080400,len:5828
[15:57:04]entry 0x400806a8
[15:57:05]
[15:57:05]00:00:00.002 HDW: ESP32-D0WDQ6 
[15:57:05]00:00:00.008 UFS: FlashFS mounted with 312 kB free
[15:57:05]00:00:00.015 FRC: Some settings have been reset (9)
[15:57:05]Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
[15:57:05]
[15:57:05]Core  1 register dump:
[15:57:05]PC      : 0x4011498f  PS      : 0x00060c30  A0      : 0x8010497e  A1      : 0x3ffb2700  
[15:57:05]A2      : 0x807a6030  A3      : 0x3ffc8784  A4      : 0x00000002  A5      : 0x00000000  
[15:57:05]A6      : 0x3ffc8200  A7      : 0x3ffb2700  A8      : 0x800e2bea  A9      : 0x3ffb2730  
[15:57:05]A10     : 0x00000045  A11     : 0x00000000  A12     : 0x00000045  A13     : 0x00004080  
[15:57:05]A14     : 0x00000001  A15     : 0x3ffb08c0  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c  
[15:57:05]EXCVADDR: 0x0000000c  LBEG    : 0x400870fd  LEND    : 0x40087105  LCOUNT  : 0x00000027  
[15:57:05]
[15:57:05]
[15:57:05]Backtrace:0x4011498c:0x3ffb27000x4010497b:0x3ffb2770 0x40150351:0x3ffb2790 0x400e601e:0x3ffb2820 
[15:57:05]
[15:57:05]
[15:57:05]
[15:57:05]
[15:57:05]ELF file SHA256: 0000000000000000
[15:57:05]
[15:57:05]Rebooting...
[15:57:05]ets Jun  8 2016 00:22:57
[15:57:05]
[15:57:05]rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
[15:57:05]configsip: 0, SPIWP:0xee
[15:57:05]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[15:57:05]mode:DOUT, clock div:2
[15:57:05]load:0x3fff0018,len:4
[15:57:05]load:0x3fff001c,len:1044
[15:57:05]load:0x40078000,len:10124
[15:57:05]load:0x40080400,len:5828
[15:57:05]entry 0x400806a8
Serial port closed!
~~~

**Comment by trlafleur** [comment]
ok, found the issue, I was using a variable that was in the struct prior to allocating the struct... a lot of effort to save 92 bytes when in the esp32, we still have over 263k or ram remaining... but, a good learning experience...

thanks...

**Reply by barbudor** [reply]
The point is more that if you load a lot of drivers in the flash (like tasmota-mega or tasmota-platinum), if each driver is assigning memory statically while not used, this is wasted and may prevent other features to work well.
Every byte count 💧 😂

**Comment by trlafleur** [comment]
What causes the Settings (TSettings) struct to be saved?? 
it's updated at each new code build, and at some rate??

**Reply by arendst** [reply]
Every secomd a hash is calculated over TSettings. If different it is saved. Simple. Your code doesn't need to know this.

In adition it is saved before every restart.

**Reply by sfromis** [reply]
Using the `SaveData` command, you can control how quickly changes are persisted to flash

**Comment by trlafleur** [comment]
Thanks, That's what I expected....

**Comment by trlafleur** [comment]
I want to store defaults in Settings.h at the first startup of new code... how can I tell if 1st startup or a reboot

**Reply by barbudor** [reply]
In `Settings.ino`, `void SettingsDefault(void)` is called whenever the Settings in flash are clear (1st boot or reset 1/2)
There are 3 sub functions `SettingsDefaultSet1..3()` for different parts of the Settings structure.
Default settings should, as much as possible, be initialized from an overridable define so anyone could hard-code in its `user_config_override.h` it's own default value.

For example, `flag3.pcf8574_ports_inverted` which is for the xdrv_28_pcf8574 driver
In `void SettingsDefaultSet2(void)` sets the default value for the flag on 1st boot based on `PCF8574_INVERT_PORTS`
```
1217: flag3.pcf8574_ports_inverted |= PCF8574_INVERT_PORTS;
```
Default value for `PCF8574_INVERT_PORTS`is defined in `my_user_config.h`
```
355: #define PCF8574_INVERT_PORTS   false  
```
Which can be overriden by anyone in `user_config_override.h` using
```
#ifdef PCF8574_INVERT_PORTS   
#undef PCF8574_INVERT_PORTS   
#endif
#define PCF8574_INVERT_PORTS   true
```

**Comment by trlafleur** [comment]
Thanks for the help...  You may want to look over what I have done, to see if I have complied with the needed Tasmota standards. If you find my project of interest, I could do a pull request in the future. At this point, I have only tested with ESP32, and ESP32S3.

I have 8 Tasmota devices running with this code connected to water meters and flow sensors.

**Comment by trlafleur** [comment]
I have updated my code base to ver 1.4 under Tasmota 12.3.1.1
settings are now saved in flash and not in settings.h

~~~
https://github.com/trlafleur/Flow-Sensor-Tasmota
~~~

**Comment by Moominza** [comment]
hello, i am struggling to get into esp32. the device keeps on trying to connect to Lafleur AP. if i reset esp32 it just load it again from setting file  "Settings loaded from file: /.snsset125" am i suppose to set up my network another way?
![image](https://github.com/user-attachments/assets/b88933db-d138-409e-93d6-203a7e0757b9)

![image](https://github.com/user-attachments/assets/1e757428-391a-4ffd-a5c7-5530c3b01338)

**Reply by trlafleur** [reply]
I'm on travel, will not be able to look at this until I return....
You should be able to re-built from source instead of pre-built binary's...
You should be able to change SSID via commands via USB port..

**Reply by Moominza** [reply]
I am Have been struggling to rebuild the binary. i was able to set my SSID the only issue that I really have left is to customize the parameters to metric. is it possible to change it without rebuilding?

**Comment by trlafleur** [comment]
I’m in Africa, back home end of September

There are command via serial or web to change everything




~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~


Tom Lafleur

**Comment by Moominza** [comment]
Not south africa by chance? Thank you.


On Sun, 1 Sept 2024 at 18:33, Tom Lafleur ***@***.***> wrote:

> I’m in Africa, back home end of September
>
> There are command via serial or web to change everything
>
>
>
>
> ~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~
>
>
> Tom Lafleur
>
>
>
> On Sun, Sep 1, 2024 at 1:10 PM sfromis ***@***.***> wrote:
>
> > Details depend on what you specifically mean by "customize the paramters
> > to metric", as defaults are already metric. For temperature there is
> > SetOption8 with SetOption24 for atmospheric pressure.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/arendst/Tasmota/discussions/15302#discussioncomment-10511660>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABC4EK6RFHHN6BDKE3JSUGDZULY4HAVCNFSM6AAAAABL5YR6NCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJRGE3DMMA>
>
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/15302#discussioncomment-10513284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE32SYVMB37JIM642APDMSLZUM6T5AVCNFSM6AAAAABL5YR6NCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJRGMZDQNA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Etienne Fourie

**Comment by trlafleur** [comment]
Yes, I’m Zimbabwe


~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur
